packages/hardhat-core/src/internal/artifacts.ts

```diff
@@ -40,6 +41,24 @@ import {
 
 const log = debug("hardhat:core:artifacts");
 
+interface IHardhatArtifactSource extends ArtifactSource {
```

Comment:
It seems like we have three options here:
- The core logic is just another ArtifactSource and we are dogfooding the interface. Initially I though this was an interesting idea, but now I'm not sure if it's easily doable. Plus, we probably want to have "special powers" here. And it's not like we are going to have dozens of implementations of the interface (unlike the task system, where the fact that we are dogfooding it does seem valuable).
- The core logic for getting the artifacts is "an ArtifactSource just like any other, but not really", which seems to be the current approach.
- The core logic is part of the Artifacts implementation itself and we don't pretend it's not special.

The three options seem sensible, but I think I'd just go with the third one just because it's straightforward and the benefits of the other more indirect options are not super clear to me.

Reply:
The existing `Artifacts` class is quite tangled, and I opted for the dogfooding approach primarily as a means of verifying the new logic to use `ArtifactSources`. I intend to make some further refactorings to that class (likely separating it into layers), but i felt like this was a stable intermediate step to share for feedback.

Also, note that this current approach does support "special powers", via special-case delegation to only `_sourcesInPriorityOrder[0]` in the `Artifacts` methods that don't exist on `ArtifactSource`s, or, more specifically, those methods listed in this interface you commented on.

To me, the biggest benefit of the indirect options is to avoid growing the existing `Artifacts` class. Yes, it would be straightforward, I think we should attempt to take a step in the direction of breaking that class apart rather than growing it, and this separation of `Source` from `SourceManager` was the best thing I could think of.